### PR TITLE
Jj adapt correct ctf to diff sizes

### PR DESCRIPTION
**Pull request by JorMaister**
- Refactorization: avoid the excesive usage of getFirstItem.
  - File generation methods (most of them in utils) adapted to the CTF-TS intersection functionality.

  - Protocol CTF correction:
    * Now it can deal with excluded views in the CTF and/or in the TS.
    * The xf, defocus, and tlt files are generated inside each TS extra sub-directory.

All the combinations of TS and CTF states working fine in protocol CTF correction:

Left column: TS state
Right column: CTF state

States:
- Excluded views: views excluded at metadata level (from user's actions or from some protocols, like aretomo)
- Re-stacked: views excluded at stack level, e. g., some interpolated data
- Unmodified: no excluded views nor re-stacked

Unmodified | Unmodified
-- | --
Unmodified | Excluded Views
Unmodified | Re-stacked
Excluded Views | Unmodified
Excluded Views | Excluded Views
Excluded Views | Re-stacked
Re-stacked | Unmodified
Re-stacked | Excluded Views
Re-stacked | Re-stacked

### CODE REVIEW

File: imod/protocols/protocol_base.py [hunk 1]
```diff
@@ -229,95 +241,139 @@ def getTmpTSFile(self, tsId, tmpPrefix=None, suffix=".mrcs"):
...
         return os.path.join(tmpPrefix, tsId + suffix)
 
-    def convertInputStep(self, tsObjId, generateAngleFile=True,
-                         imodInterpolation=True, doSwap=False, oddEven=False):
+    def genTsPaths(self, tsId):
+        """Generate the subdirectories corresponding to the current tilt-series in tmp and extra"""
+        path.makePath(*[self._getExtraPath(tsId), self._getTmpPath(tsId)])
+
+    @staticmethod
+    def getOutTsFileName(tsId, suffix=None, ext='st'):
```
Comment (azazellochg) [inline imod/protocols/protocol_base.py:249]: can we avoid using ".st" extensions? IMOD works nicely with mrc/mrcs.

File: imod/protocols/protocol_ctfEstimation_automatic.py [hunk 2]
```diff
@@ -434,21 +435,17 @@ def ctfEstimation(self, tsObjId, expDefoci):
...
-    def createOutputStep(self, tsObjId, outputSetName=OUTPUT_CTF_SERIE):
-        ts = self._getTiltSeries(tsObjId)
-        tsId = ts.getTsId()
-
-        extraPrefix = self._getExtraPath(tsId)
-        defocusFilePath = os.path.join(extraPrefix,
-                                       ts.getFirstItem().parseFileName(extension=".defocus"))
+    def createOutputStep(self, tsId, outputSetName=OUTPUT_CTF_SERIE):
+        ts = self.tsDict[tsId]
+        defocusFilePath = self.getExtraOutFile(tsId, ext="defocus")
```
Comment (azazellochg) [inline imod/protocols/protocol_ctfEstimation_automatic.py:440]: use DEFOCUS_EXT?

File: imod/protocols/protocol_fiducialModel.py [hunk 3]
```diff
@@ -502,28 +445,21 @@ def xcorrStep(self, tsObjId):
...
-
         BordersInXandY = '%d,%d' % (borders[0], borders[1])
-        SizeOfPatchesXandY = '%d,%d'  % (sizePatches[0], sizePatches[1])
+        SizeOfPatchesXandY = '%d,%d' % (sizePatches[0], sizePatches[1])
 
         paramsTiltXCorr = {
-            'inputFile': os.path.join(tmpPrefix, firstItem.parseFileName()),
-            'outputFile': os.path.join(extraPrefix, firstItem.parseFileName(suffix="_pt", extension=".fid")),
+            'inputFile': self.getTmpOutFile(tsId),
+            'outputFile': self.getExtraOutFile(tsId, suffix="pt", ext=".fid"),
```
Comment (azazellochg) [inline imod/protocols/protocol_fiducialModel.py:462]: ext="fid" instead?

File: imod/protocols/protocol_fiducialModel.py [hunk 4]
```diff
@@ -560,27 +496,21 @@ def xcorrStep(self, tsObjId):
...
-        firstItem = ts.getFirstItem()
+        """
         MinimumOverlap = 4
         AssignSurfaces = 1
         LengthOfPieces = -1
 
         paramschopconts = {
-            'inputFile': os.path.join(extraPrefix, firstItem.parseFileName(suffix="_pt", extension=".fid")),
-            'outputFile': os.path.join(extraPrefix, firstItem.parseFileName(suffix="_gaps", extension=".fid")),
+            'inputFile': self.getExtraOutFile(tsId, suffix="pt", ext=".fid"),
```
Comment (azazellochg) [inline imod/protocols/protocol_fiducialModel.py:512]: ext="fid" instead?

File: imod/protocols/protocol_fiducialModel.py [hunk 5]
```diff
@@ -560,27 +496,21 @@ def xcorrStep(self, tsObjId):
...
+            'outputFile': self.getExtraOutFile(tsId, suffix="gaps", ext=".fid"),
```
Comment (azazellochg) [inline imod/protocols/protocol_fiducialModel.py:513]: ext="fid" instead?

**Review by azazellochg — COMMENTED** [review]
Very nice work @JorMaister ! I like tsDict and filenames handling improvements. 
I went once through the changes and will make a few more passes before approving.

File: imod/utils.py [hunk 6]
```diff
@@ -818,30 +832,30 @@ def generateDefocusIMODFileFromObject(ctfTomoSeries, defocusFilePath,
...
-                    # CONVERT DEFOCUS VALUE TO NANOMETERS (IMOD CONVENTION)
-                    ctfTomo.getDefocusU()/10,
-                    # CONVERT DEFOCUS VALUE TO NANOMETERS (IMOD CONVENTION)
-                    ctfTomo.getDefocusV()/10,
-                    ctfTomo.getDefocusAngle())
-                           )
-
-                lines.append(newLine)
+                    lines.append(newLine)
+                    ind += 1
```
[the remark this answers is not in the document]
Comment (JorMaister) [reply imod/utils.py:858]: Thanks, @azazellochg.
Yes, it only has to be updated if there's a CTF associated with the current tilt-image, so the defocus files is generated containing only the present views.

File: imod/protocols/protocol_fiducialModel.py (hunk 5 above)
Comment (JorMaister) [reply imod/protocols/protocol_fiducialModel.py:513]: Thanks, @azazellochg,

I've corrected those and made some more extension centralization.

**Review by azazellochg — APPROVED** [review]
The tests are passing. I am not able to check patch tracking until the weekend, if any issues arise we can fix them later

**Pull request merged**